myia_backend_python/myia_backend_python/python.py

```diff
@@ -462,6 +462,82 @@ def nested_list_to_code_string(structure, indentation=""):
     return code
 
 
+class PdbCall:
+    """Helper class to run generated code with PDB.
+
+    Compiled code contains only closures and main function, but a call
+    must provide main function inputs, which may be integers or
+    numpy arrays. To deal with it, this class uses pickle to dump/load
+    main inputs/output, and extends generated code to produce a valid
+    script which will handle pickled data from temporary files.
+    """
+
+    def __init__(self, code, pdb_fn_name):
+        """ Initialize.
+
+        :param code: generated code
+        :param pdb_fn_name: a name for the main function that will be
+            called from generated script.
+        """
+        self.code = code
+        self.pdb_fn_name = pdb_fn_name
+
+    def __call__(self, *args):
+        """Execute main function with given args."""
+        import os
+        import pickle
+        import subprocess
+        import tempfile
+
+        # File for generated script.
+        code_fd, code_path = tempfile.mkstemp(
+            prefix="myia_backend_python_code_", suffix=".py"
+        )
+        # File to pickle main function args.
+        args_fd, args_path = tempfile.mkstemp()
+        # File to pickle main function return value,
+        out_fd, out_path = tempfile.mkstemp()
+        # Extension code to call main function.
+        # Load main function args from args_path.
+        # Save main function output into out_path.
+        pdb_code = f"""
+def {self.pdb_fn_name}():
+    import pickle
+    with open("{args_path}", "rb") as input_file:
+        main_parameters = pickle.load(input_file)
+    main_output = main(*main_parameters)
+    with open("{out_path}", "wb") as output_file:
+        pickle.dump(main_output, output_file)
+
+if __name__ == "__main__":
+    {self.pdb_fn_name}()
+"""
+        # Generate script code.
+        final_code = self.code + pdb_code
+        try:
+            # Save script.
+            with open(code_path, "w") as code_file:
+                code_file.write(final_code)
+            # Save main args.
+            with open(args_path, "wb") as input_file:
+                pickle.dump(args, input_file)
+            # Run script.
+            subprocess.run(["python", "-m", "pdb", code_path])
```

Comment:
This will start debugging at the entry of the module you are generating, which isn't super useful: it will start at `import math` and you'll have to do next a bunch of times and then step inside `main`. Ideally we want to start debugging inside `main` directly.

I think a better way would be something like this:

```python
# Create a globals dictionary
glb = {}

# Execute the code for the function using glb as the globals
exec(self.code, glb)

# Run pdb at the call, assuming the main function is called "main"
pdb.runcall(glb["main"], *args)
```

I'm not sure how to write a test for that, but it's somewhat less important to test debug stuff, since you're not going to run it in production.